src/credentials/StaticGroupDataProvider.cpp

```diff
@@ -0,0 +1,514 @@
+                return CHIP_NO_ERROR;
+            }
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR RemoveGroupAllMappings(chip::FabricIndex fabric_index, EndpointId endpoint) override
+    {
+        Fabric * fabric = GetFabric(fabric_index);
+        // Remove all mappings from fabric
+        for (uint16_t i = 0; fabric && i < kEndpointEntriesMax; ++i)
+        {
+            fabric->endpoints[i].in_use = false;
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    GroupMappingIterator * IterateGroupMappings(chip::FabricIndex fabric_index, EndpointId endpoint) override
+    {
+        return new EndpointIterator(GetFabric(fabric_index), endpoint);
+    }
+
+    // States
+    CHIP_ERROR SetGroupState(chip::FabricIndex fabric_index, GroupState & state) override
+    {
+        Fabric * fabric    = GetFabric(fabric_index);
+        StateEntry * entry = nullptr;
+
+        // Search for existing, or unused entry
+        for (uint16_t i = 0; fabric && i < kStateEntriesMax; ++i)
+        {
+            if (fabric->states[i].in_use)
+            {
+                if (fabric->states[i].state_index == state.state_index)
+                {
+                    // Reuse existing entry
+                    entry = &fabric->states[i];
+                    break;
+                }
+            }
+            else if (!entry)
+            {
+                // Unused entry
+                entry = &fabric->states[i];
+            }
+        }
+        if (entry)
+        {
+            entry->state_index   = state.state_index;
+            entry->group         = state.group;
+            entry->key_set_index = state.key_set_index;
+            entry->in_use        = true;
+            return CHIP_NO_ERROR;
+        }
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    CHIP_ERROR GetGroupState(chip::FabricIndex fabric_index, GroupState & state) override
+    {
+        Fabric * fabric = GetFabric(fabric_index);
+        // Search for existing state
+        for (uint16_t i = 0; fabric && i < kStateEntriesMax; ++i)
+        {
+            StateEntry & entry = fabric->states[i];
+            if (entry.in_use && entry.state_index == state.state_index)
+            {
+                // Found
+                state.group         = entry.group;
+                state.key_set_index = entry.key_set_index;
+                return CHIP_NO_ERROR;
+            }
+        }
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+
+    CHIP_ERROR RemoveGroupState(chip::FabricIndex fabric_index, uint16_t state_index) override
+    {
+        Fabric * fabric = GetFabric(fabric_index);
+        // Search for existing state
+        for (uint16_t i = 0; fabric && i < kStateEntriesMax; ++i)
+        {
+            StateEntry & entry = fabric->states[i];
+            if (entry.in_use && entry.state_index == state_index)
+            {
+                // Found
+                entry.in_use = false;
+                return CHIP_NO_ERROR;
+            }
```

Comment:
On removal of a group state, we should likely have a `GarbageCollectGroupsWithoutKeys` to affect the following part of the spec:

> [[ref_cmd_KeySetRemove,KeySetRemove]]
> ===== KeySetRemove Command
> 
> ....
> 
> ====== Effect on Receipt
> If there exists a Group Key Set associated with the accessing fabric which has the same `GroupKeySetIndex` as that provided in the `GroupKeySetIndex` field, then the contents of that Group Key Set SHALL be removed, including all epoch keys it contains.
> 
> If there exist any entries for the accessing fabric within the <<ref_GroupKeyManagementCluster_GroupsAttribute,Groups>> attribute that refer to the `GroupKeySetIndex` just removed, then these entries SHALL be removed from that list.
> 
> If there exist any entries for the accessing fabric within the <<ref_GroupKeyManagementCluster_GroupsAttribute,Groups>> attribute that refer to the `GroupKeySetIndex` just removed, then these entries SHALL be removed from that list.
>